Firestore/core/test/firebase/firestore/local/mutation_queue_test.cc

```diff
@@ -0,0 +1,432 @@
+    MutationBatch batch = AddMutationBatch();
+    batches.push_back(batch);
+  }
+
+  return batches;
+}
+
+/** Returns the number of mutation batches in the mutation queue. */
+size_t MutationQueueTestBase::BatchCount() {
+  return mutation_queue_->AllMutationBatches().size();
+}
+
+/**
+ * Removes the first n entries from the the given batches and returns them.
+ *
+ * @param n The number of batches to remove.
+ * @param batches The array to mutate, removing entries from it.
+ * @return A new array containing all the entries that were removed from @a
+ * batches.
+ */
+std::vector<MutationBatch> MutationQueueTestBase::RemoveFirstBatches(
+    size_t n, std::vector<MutationBatch>* batches) {
+  std::vector<MutationBatch> removed(batches->begin(), batches->begin() + n);
+  batches->erase(batches->begin(), batches->begin() + n);
+
+  for (const MutationBatch& batch : removed) {
+    mutation_queue_->RemoveMutationBatch(batch);
+  }
+  return removed;
+}
+
+MutationQueueTest::MutationQueueTest() : MutationQueueTestBase(GetParam()()) {
+}
+
+TEST_P(MutationQueueTest, CountBatches) {
+  persistence_->Run("test_count_batches", [&] {
+    ASSERT_EQ(0, BatchCount());
+    ASSERT_TRUE(mutation_queue_->IsEmpty());
+
+    MutationBatch batch1 = AddMutationBatch();
+    ASSERT_EQ(1, BatchCount());
+    ASSERT_FALSE(mutation_queue_->IsEmpty());
+
+    MutationBatch batch2 = AddMutationBatch();
+    ASSERT_EQ(2, BatchCount());
+
+    mutation_queue_->RemoveMutationBatch(batch1);
+    ASSERT_EQ(1, BatchCount());
+
+    mutation_queue_->RemoveMutationBatch(batch2);
+    ASSERT_EQ(0, BatchCount());
+    ASSERT_TRUE(mutation_queue_->IsEmpty());
+  });
+}
+
+TEST_P(MutationQueueTest, AcknowledgeBatchID) {
+  persistence_->Run("test_acknowledge_batch_id", [&] {
+    ASSERT_EQ(BatchCount(), 0);
+
+    MutationBatch batch1 = AddMutationBatch();
+    MutationBatch batch2 = AddMutationBatch();
+    MutationBatch batch3 = AddMutationBatch();
+    ASSERT_GT(batch1.batch_id(), kBatchIdUnknown);
+    ASSERT_GT(batch2.batch_id(), batch1.batch_id());
+    ASSERT_GT(batch3.batch_id(), batch2.batch_id());
+
+    ASSERT_EQ(BatchCount(), 3);
+
+    mutation_queue_->AcknowledgeBatch(batch1, {});
+    mutation_queue_->RemoveMutationBatch(batch1);
+    ASSERT_EQ(BatchCount(), 2);
+
+    mutation_queue_->AcknowledgeBatch(batch2, {});
+    ASSERT_EQ(BatchCount(), 2);
+
+    mutation_queue_->RemoveMutationBatch(batch2);
+    ASSERT_EQ(BatchCount(), 1);
+
+    mutation_queue_->RemoveMutationBatch(batch3);
+    ASSERT_EQ(BatchCount(), 0);
+  });
+}
+
+TEST_P(MutationQueueTest, AcknowledgeThenRemove) {
+  persistence_->Run("test_acknowledge_then_remove", [&] {
+    MutationBatch batch1 = AddMutationBatch();
+
+    mutation_queue_->AcknowledgeBatch(batch1, {});
+    mutation_queue_->RemoveMutationBatch(batch1);
+
+    ASSERT_EQ(BatchCount(), 0);
+  });
+}
+
+TEST_P(MutationQueueTest, LookupMutationBatch) {
+  // Searching on an empty queue should not find a non-existent batch
```

Comment:
Should this comment be moved inside the lambda?

Reply:
Done.